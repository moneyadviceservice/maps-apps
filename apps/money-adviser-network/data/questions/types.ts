export const FORM_FIELDS = {
  consentDetails: 'consentDetails',
  consentReferral: 'consentReferral',
  consentOnline: 'consentOnline',
  customerReference: 'customerReference',
  departmentName: 'departmentName',
  whenToSpeak: 'whenToSpeak',
  timeSlot: 'timeSlot',
  firstName: 'firstName',
  lastName: 'lastName',
  telephone: 'telephone',
  email: 'email',
  postcode: 'postcode',
  securityQuestion: 'securityQuestion',
  securityAnswer: 'securityAnswer',
} as const;

export const SingleFieldGroups = {
  consentDetails: ['consentDetails'],
  consentReferral: ['consentReferral'],
  consentOnline: ['consentOnline'],
  whenToSpeak: ['whenToSpeak'],
  timeSlot: ['timeSlot'],
} as const;

const MultiFieldGroups = {
  reference: ['customerReference', 'departmentName'],
  customerDetails: ['firstName', 'lastName', 'telephone', 'email'],
  securityQuestions: ['postcode', 'securityQuestion', 'securityAnswer'],
} as const;

const FIELD_GROUPS = {
  ...SingleFieldGroups,
  ...MultiFieldGroups,
} as const;

export const FORM_GROUPS = Object.keys(FIELD_GROUPS).reduce((acc, key) => {
  acc[key as keyof typeof FIELD_GROUPS] = key as keyof typeof FIELD_GROUPS;
  return acc;
}, {} as Record<keyof typeof FIELD_GROUPS, keyof typeof FIELD_GROUPS>);

export type Fields = {
  [Key in keyof typeof FORM_FIELDS]: string;
};

export type CookieData = {
  [Group in keyof typeof SingleFieldGroups]?: {
    value: string;
  };
} & {
  [Group in keyof typeof MultiFieldGroups]?: {
    [Field in (typeof MultiFieldGroups)[Group][number]]: string;
  };
};
