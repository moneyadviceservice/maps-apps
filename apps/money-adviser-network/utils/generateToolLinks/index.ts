export { generateToolLinks } from './generateToolLinks';

export type { ToolLinks } from './generateToolLinks';
