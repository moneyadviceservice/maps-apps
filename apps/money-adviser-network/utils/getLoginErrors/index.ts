export { getLoginAcdlErrors, getLoginErrors } from './getLoginErrors';
