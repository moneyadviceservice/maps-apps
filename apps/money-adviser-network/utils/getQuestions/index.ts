export { FLOW, getQuestions } from './getQuestions';
