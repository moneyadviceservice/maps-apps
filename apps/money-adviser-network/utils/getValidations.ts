const postcodeRegex = /^[A-Z]{1,2}\d[A-Z\d]? ?\d[A-Z]{2}$/i;

export const validatePostcode = (value: string) => !postcodeRegex.test(value);
