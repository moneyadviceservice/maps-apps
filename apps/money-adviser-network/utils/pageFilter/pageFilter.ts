import getConfig from 'next/config';

import { PAGES } from 'CONSTANTS';
import { ParsedUrlQuery } from 'querystring';

import { addEmbedQuery } from '@maps-react/utils/addEmbedQuery';

import { getCurrentPath } from '../../utils/getCurrentPath';
import { FLOW } from '../../utils/getQuestions';

export type RecordValue = any;

export type DataFromQuery = Record<string, RecordValue>;

const QUESTION_PREFIX = 'q-';
const CHANGE_ANSWER_PARAM = 'changeAnswer';

export interface PageFilterFunctions {
  getDataFromQuery: () => DataFromQuery;
  convertQueryParamsToString: (
    p: Record<string, RecordValue>,
  ) => URLSearchParams;
  goToStep: (step: string) => string;
  goToFirstStep: () => string;
  backStep: (
    currentStep: number,
    storedData: Record<string, RecordValue>,
    cookieData: Record<string, RecordValue>,
    flow: FLOW,
    url: string | undefined,
  ) => string;
  goToResultPage: () => string;
  goToLastQuestion: (data: object, flow: FLOW) => string;
  goToChangeOptionsPage: () => string;
  goToQuestion: (question: number) => string;
}

export const pageFilter = (
  query: ParsedUrlQuery,
  path: string,
  isEmbed: boolean,
  prodLandingPageLink?: string,
  pagePrefix?: string,
): PageFilterFunctions => {
  const {
    publicRuntimeConfig: { ENVIRONMENT },
  } = getConfig();

  const lang = query?.language;
  const prefix = pagePrefix ?? 'question';

  const getDataFromQuery = () => {
    const data: DataFromQuery = {};
    Object.keys(query).forEach((g) => {
      if (
        g.indexOf(QUESTION_PREFIX) > -1 ||
        g.indexOf(CHANGE_ANSWER_PARAM) > -1 ||
        g.indexOf('error') > -1
      ) {
        data[g] = query[g];
      }
    });

    return data;
  };

  const convertQueryParamsToString = (
    p: Record<string, RecordValue>,
  ): URLSearchParams => {
    return new URLSearchParams(p as Record<string, string>);
  };

  return {
    getDataFromQuery,
    convertQueryParamsToString,
    goToStep: (step: string) =>
      `/${lang}${path}${step}?${convertQueryParamsToString(
        getDataFromQuery(),
      )}${addEmbedQuery(isEmbed, '&')}`,
    goToFirstStep: () =>
      `/${lang}${path}${prefix}-1?restart=true${addEmbedQuery(isEmbed, '&')}`,
    backStep: (
      currentStep: number,
      storedData: Record<string, RecordValue>,
      cookieData: Record<string, RecordValue>,
      flow: FLOW,
      url: string | undefined,
    ) => {
      if (currentStep === 1 && flow === FLOW.START) {
        const backUrl = `/${lang}${path}${addEmbedQuery(isEmbed, '?')}`;
        if (ENVIRONMENT === 'production') {
          return prodLandingPageLink ?? backUrl;
        }
        return backUrl;
      }

      if (url?.includes(PAGES.DEBT_ADVICE_LOCATOR) && flow === FLOW.ONLINE) {
        return `/${lang}/${getCurrentPath(
          FLOW.START,
        )}/${prefix}-4?${convertQueryParamsToString(
          getDataFromQuery(),
        )}${addEmbedQuery(isEmbed, '&')}`;
      }

      if (currentStep === 1 && flow !== FLOW.START) {
        const lastQuestion = Object.keys(storedData)
          .filter((key) => key !== 'error')
          .reduce((acc, curr) => curr, '');

        return `/${lang}/money-adviser-network/${
          FLOW.START
        }/${lastQuestion}?${convertQueryParamsToString(
          getDataFromQuery(),
        )}${addEmbedQuery(isEmbed, '&')}`;
      }
      const previousStep = currentStep - 1;

      if (flow === FLOW.ONLINE || flow === FLOW.TELEPHONE) {
        if (url?.includes(PAGES.CONSENT_REJECTED)) {
          return `/${lang}${path}${prefix}-1?${convertQueryParamsToString(
            getDataFromQuery(),
          )}${addEmbedQuery(isEmbed, '&')}`;
        }
      }

      if (flow === FLOW.TELEPHONE) {
        if (
          (currentStep === 4 &&
            cookieData['consentReferral']?.consentReferralAnswer === '1') ||
          (currentStep === 6 && cookieData['t-4'] === '0')
        ) {
          return `/${lang}${path}${prefix}-${
            previousStep - 1
          }?${convertQueryParamsToString(getDataFromQuery())}${addEmbedQuery(
            isEmbed,
            '&',
          )}`;
        }
      }

      return previousStep > -1
        ? `/${lang}${path}${prefix}-${previousStep}?${convertQueryParamsToString(
            getDataFromQuery(),
          )}${addEmbedQuery(isEmbed, '&')}`
        : `/${lang}${path}${prefix}-${currentStep}?${convertQueryParamsToString(
            getDataFromQuery(),
          )}${addEmbedQuery(isEmbed, '&')}`;
    },
    goToResultPage: () =>
      `/${lang}${path}results?${convertQueryParamsToString(
        getDataFromQuery(),
      )}${addEmbedQuery(isEmbed, '&')}`,
    goToLastQuestion: (data: object, flow: FLOW) => {
      let lastStep;
      if (flow === FLOW.TELEPHONE) {
        lastStep = 7;
      } else if (flow === FLOW.ONLINE) {
        lastStep = 3;
      } else {
        lastStep = Object.keys(data).reduce((a, c) => {
          const val = c.split(QUESTION_PREFIX);
          if (val.length > 1) {
            return a < Number(val[1]) ? Number(val[1]) : a;
          }
          return a;
        }, 0);
      }

      return `/${lang}${path}${prefix}-${lastStep}?${convertQueryParamsToString(
        getDataFromQuery(),
      )}${addEmbedQuery(isEmbed, '&')}`;
    },
    goToChangeOptionsPage: () =>
      `/${lang}${path}change-options?${convertQueryParamsToString(
        getDataFromQuery(),
      )}${addEmbedQuery(isEmbed, '&')}`,
    goToQuestion: (question: number) => {
      return `/${lang}${path}${prefix}-${question}?${convertQueryParamsToString(
        getDataFromQuery(),
      )}${addEmbedQuery(isEmbed, '&')}`;
    },
  };
};
