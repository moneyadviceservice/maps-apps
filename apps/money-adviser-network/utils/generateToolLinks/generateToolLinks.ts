import { FLOW } from '../../utils/getQuestions';
import { DataFromQuery, PageFilterFunctions } from '../../utils/pageFilter';

export type ToolLinks = {
  question: {
    backLink: string;
    goToQuestionOne: string;
    goToQuestionTwo: string;
    goToQuestionThree: string;
  };
  change: {
    backLink: string;
    nextLink: string;
  };
  confirmation: {
    backLink: string;
    start: string;
  };
};

export const generateToolLinks = (
  saveddata: DataFromQuery,
  cookieData: DataFromQuery,
  filter: PageFilterFunctions,
  currentStep: number,
  currentFlow: FLOW,
  url: string | undefined,
): ToolLinks => {
  const checkFlow = () => {
    return currentFlow !== FLOW.START
      ? filter.backStep(currentStep, saveddata, cookieData, currentFlow, url)
      : '';
  };
  return {
    question: {
      backLink:
        currentStep > 1
          ? filter.backStep(
              currentStep,
              saveddata,
              cookieData,
              currentFlow,
              url,
            )
          : checkFlow(),
      goToQuestionOne: filter.goToQuestion(1),
      goToQuestionTwo: filter.goToQuestion(2),
      goToQuestionThree: filter.goToQuestion(3),
    },
    change: {
      backLink: filter.goToLastQuestion(saveddata, currentFlow),
      nextLink: filter.goToResultPage(),
    },
    confirmation: {
      backLink: filter.goToChangeOptionsPage(),
      start: filter.goToFirstStep(),
    },
  };
};
