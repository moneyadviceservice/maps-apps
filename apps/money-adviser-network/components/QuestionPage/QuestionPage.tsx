import { ReactNode } from 'react';
import { DataFromQuery } from '@maps-react/utils/pageFilter';
import { ToolLinks } from 'utils/generateToolLinks';
import { Questions } from '@maps-react/form/components/Questions';
import { ErrorType, Question } from '@maps-react/form/types';

type Props = {
  storedData: DataFromQuery;
  data: string;
  currentStep: number;
  links: ToolLinks;
  isEmbed: boolean;
  toolPath: string;
  questions: Question[];
  errors: ErrorType[];
  helpSection?: ReactNode;
  pageError?: ReactNode;
};

export const QuestionPage = ({
  storedData,
  data,
  currentStep,
  links,
  isEmbed,
  toolPath,
  questions,
  errors,
  helpSection,
  pageError,
}: Props) => (
  <Questions
    storedData={storedData}
    data={data}
    questions={questions}
    errors={errors}
    currentStep={currentStep}
    backLink={links.question.backLink}
    dataPath={toolPath}
    apiCall={'/api/submit-answer'}
    isEmbed={isEmbed}
    alwaysDisplaySubText={true}
    bottomInfo={helpSection}
    topInfo={pageError}
  />
);
