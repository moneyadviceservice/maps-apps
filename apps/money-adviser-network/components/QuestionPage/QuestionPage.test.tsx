import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { QuestionPage } from './QuestionPage';
import { Questions } from '@maps-react/form/components/Questions';
import { ToolLinks } from 'utils/generateToolLinks';
import { DataFromQuery } from '@maps-react/utils/pageFilter';
import { Question, Answer } from '@maps-react/form/types';

jest.mock('@maps-react/form/components/Questions', () => ({
  Questions: jest.fn(() => <div>Mocked Questions Component</div>),
}));

describe('QuestionPage', () => {
  const mockStoredData: DataFromQuery = {
    someKey: 'someValue',
  };

  const mockToolLinks: ToolLinks = {
    question: {
      backLink: '/back',
      goToQuestionOne: '1',
      goToQuestionTwo: '2',
    },
    change: {
      backLink: '/back',
      nextLink: '/next',
    },
    summary: {
      backLink: '/back',
      nextLink: '/next',
    },
    result: {
      backLink: '/back',
      firstStep: '/first',
    },
  };

  const mockAnswers: Answer[] = [
    {
      text: 'Answer 1',
    },
    {
      text: 'Answer 2',
    },
  ];

  const mockQuestions: Question[] = [
    {
      questionNbr: 1,
      group: 'Group 1',
      title: 'Question 1',
      type: 'multiple-choice',
      answers: mockAnswers,
      description: 'Description for question 1',
    },
    {
      questionNbr: 2,
      group: 'Group 2',
      title: 'Question 2',
      type: 'single-choice',
      answers: mockAnswers,
    },
  ];

  const props = {
    storedData: mockStoredData,
    data: 'test-data',
    currentStep: 1,
    links: mockToolLinks,
    isEmbed: false,
    toolPath: '/tool-path',
    questions: mockQuestions,
    errors: [],
  };

  it('should render Questions component with correct props', () => {
    const { container, getByText } = render(<QuestionPage {...props} />);

    expect(getByText('Mocked Questions Component')).toBeInTheDocument();

    expect(Questions).toHaveBeenCalledWith(
      {
        storedData: props.storedData,
        data: props.data,
        questions: props.questions,
        errors: props.errors,
        currentStep: props.currentStep,
        backLink: props.links.question.backLink,
        dataPath: props.toolPath,
        apiCall: '/api/submit-answer',
        isEmbed: props.isEmbed,
        alwaysDisplaySubText: true,
      },
      {},
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
