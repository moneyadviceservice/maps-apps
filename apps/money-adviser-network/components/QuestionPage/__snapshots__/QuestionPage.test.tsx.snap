// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionPage should render Questions component with correct props 1`] = `
<div>
  Mocked Questions Component
</div>
`;
